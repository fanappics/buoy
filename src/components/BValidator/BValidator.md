Wrapper to handle validation of a group of buoy components.

```js
  let isValid = false;

  <b-validator scope="input-validator" v-model="isValid">
    <b-input type="text" id="1" label="Required Input #1" required />
    <b-input type="text" id="2" label="Required Input #2" required />
    <div v-if="!isValid">
      Both fields are required
    </div>
  </b-validator>
```