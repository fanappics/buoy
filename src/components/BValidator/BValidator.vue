<template>
  <div>
    <slot></slot>
    {{ errors }}
  </div>
</template>

<script>
import events from '../../event-bus'

export default {
  name: 'b-validator',
  props: {
    scope: {
      type: String,
      required: true
    }
  },
  inject: {
    parentScope: {
      default: null
    }
  },
  provide () {
    return {
      parentScope: this.scope
    }
  },
  created () {
    events.$on('validateAll', this.onValidateAll)
    events.$on('errorsChanged', this.onErrorsChanged)
  },
  beforeDestroy () {
    events.$off('validateAll', this.onValidateAll)
    events.$off('errorsChanged', this.onErrorsChanged)
  },
  data () {
    return {
      valid: false
    }
  },
  computed: {
    //Wrapper around valid so it is propegated through v-model, or 'public' as I've dubbed it
    publicValid: {
      get () {
        return this.valid
      },
      set (value) {
        if (this.valid !== value) {
          this.valid = value
          this.$emit('input', this.valid)
        }
      }
    }
  },
  methods: {
    onErrorsChanged (errors, field, scope) {
      if (this.scope === scope) {
        this.errors.remove(field, this.parentScope)
        errors.forEach(error => {
          if (error.field === field)
            this.errors.add({
              field: error.field,
              msg: error.msg,
              rule: error.rule,
              scope: this.parentScope
            })
        })
        this.publicValid = !this.errors.any()
        events.$emit('errorsChanged', this.errors.items, field, this.parentScope)
      }
    },
    onValidateAll (scope) {
      if (this.scope === scope || (this.parentScope && this.parentScope === scope)) {
        this.publicValid = !this.errors.any()
        events.$emit('validate', this.scope)
      }
    }
  }
}
</script>

<style scoped>
</style>
