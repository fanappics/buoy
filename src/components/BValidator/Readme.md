external b-button documentation:

```js
<b-validator scope="parent">
  <b-validator scope="child1">
    <b-input id="email" type="email" label="email" required/>
  </b-validator>

</b-validator>
```
