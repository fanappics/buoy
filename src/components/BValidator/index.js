import BValidator from './BValidator.vue'

/* istanbul ignore next */
BValidator.install = function install (Vue) {
  Vue.component(BValidator.name, BValidator)
}

export default BValidator
