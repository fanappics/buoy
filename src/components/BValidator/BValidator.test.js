import { createLocalVue, shallow } from 'vue-test-utils'
import VeeValidate from 'vee-validate'
import BValidator from './BValidator'

const localVue = createLocalVue()
localVue.use(VeeValidate)

describe('BValidator', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BValidator, {
      propsData: {
        scope: 'scope'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
