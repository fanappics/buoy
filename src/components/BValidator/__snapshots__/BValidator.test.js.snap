// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BValidator mounted html should match snapshot 1`] = `<div></div>`;
